apax/bal/feature_maps.py

```diff
@@ -69,16 +69,67 @@ def inner(ll_params):
             g_ll = unflatten_dict(g_ll)
 
             g_flat = jax.tree_map(lambda arr: jnp.reshape(arr, (-1,)), g_ll)
-            (gw, gb), _ = jax.tree_util.tree_flatten(g_flat)
+            (gb, gw), _ = jax.tree_util.tree_flatten(g_flat)
 
-            if self.use_ntk:
-                bias_factor = 0.1
-                weight_factor = jnp.sqrt(1 / gw.shape[-1])
-                g_scaled = [weight_factor * gw, bias_factor * gb]
-            else:
-                g_scaled = [gw, gb]
+            g = [gw, gb]
+            g = jnp.concatenate(g)
+
+            return g
+
+        return ll_grad
 
-            g = jnp.concatenate(g_scaled)
+
+class LastLayerForceFeatures(FeatureTransformation, extra="forbid"):
+    """
+    Model transfomration which computes the gradient of the output
+    wrt. the specified layer.
+    """
+
+    name: Literal["ll_force_feat"] = "ll_force_feat"
+    layer_name: str = "dense_2"
```

Comment:
Does `dense_2` assume that the apax model is always two layers deep?

Reply:
that's just the default, it's user supplied. It has to be one of the linear layers though the way it is currently implemented. But it's already possible (also in IPSuite) to choose an arbitrary linear layer.
Intermediate layers don't really make sense to use. I could code it in a way that always selects the last linear layer automatically, but that's pretty low on my priority list atm.

Reply:
Okay, then this is perfectly fine.